### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by magnus-madsen**
Implemented parallel rule evaluation. 

The flag `--threads x` can be used to specify the number of concurrent threads. By default this is the number of processors in the machine.

The diff is probably hard to read, but have a quick look. Don't merge, but just send it back to me :)

### CODE REVIEW

**Comment by magnus-madsen** [comment]
The solver class is probably the most interesting.

File: main/src/ca/uwaterloo/flix/language/phase/Codegen.scala [hunk 1]
```diff
@@ -1095,7 +1095,7 @@ object Codegen {
           case BinaryOperator.NotEqual => (IF_ICMPEQ, FCMPG, DCMPG, IF_ACMPEQ, IFEQ)
         }
         e1.tpe match {
-          case Type.Unit | Type.Str | Type.Enum(_, _) if o == BinaryOperator.Equal || o == BinaryOperator.NotEqual =>
+          case Type.Unit | Type.Str if o == BinaryOperator.Equal || o == BinaryOperator.NotEqual =>
             // Unit, String, and Enum can be reference compared for equality.
```
Comment (mhyee) [inline main/src/ca/uwaterloo/flix/language/phase/Codegen.scala]: Please update the comment. Not just this one, but also the method documentation.

File: main/test/ca/uwaterloo/flix/runtime/TestBackend.scala [hunk 2]
```diff
@@ -123,9 +123,11 @@ class TestBackend extends FunSuite {
     // A public Flix instance to expose the interop functions.
     val flix = createFlix()
 
+    // TODO: This should really use the default options as a starting point.
     private def createFlix(codegen: Boolean = false) = {
       val options = Options(
         debugger = Debugger.Disabled,
+        solver = ca.uwaterloo.flix.util.SolverOpts.Default,
```
Comment (mhyee) [inline main/test/ca/uwaterloo/flix/runtime/TestBackend.scala]: I feel like you could just write `SolverOpts.Default` and clean up the imports a bit? But I didn't look too closely.

**Comment by mhyee** [comment]
Just some minor comments. I only skimmed the Solver class.

File: main/src/ca/uwaterloo/flix/runtime/Solver.scala [hunk 3]
```diff
@@ -18,40 +18,90 @@ package ca.uwaterloo.flix.runtime
...
+  type WorkList = mutable.ArrayStack[(Rule, Env)]
 
-/**
-  * A solver based on semi-naive evaluation.
-  */
-class Solver(implicit val sCtx: Solver.SolverContext) {
+  /**
+    * The type of a (partial) interpretation.
+    *
+    * An interpretation is collection of facts.
```
Comment (mhyee) [inline main/src/ca/uwaterloo/flix/runtime/Solver.scala]: This comment isn't as clear as the other two above. What does the pair `(Symbol.TableSym, Array[AnyRef])` mean? Is a fact a collection of values (`Array[AnyRef]`) associated with a relation/lattice (`Symbol.TableSym`)?

**Pull request merged**